meta-facebook/yv4-sd/src/platform/plat_mctp.c

```diff
@@ -0,0 +1,326 @@
+{
+	k_work_submit(&send_cmd_work);
+}
+
+bool mctp_add_sel_to_ipmi(common_addsel_msg_t *sel_msg)
+{
+	CHECK_NULL_ARG_WITH_RETURN(sel_msg, false);
+
+	uint8_t system_event_record = 0x02;
+	uint8_t evt_msg_version = 0x04;
+
+	pldm_msg msg = { 0 };
+	struct mctp_to_ipmi_sel_req req = { 0 };
+
+	msg.ext_params.type = MCTP_MEDIUM_TYPE_SMBUS;
+	msg.ext_params.smbus_ext_params.addr = I2C_BUS_BMC;
+
+	msg.hdr.pldm_type = PLDM_TYPE_OEM;
+	msg.hdr.cmd = PLDM_OEM_IPMI_BRIDGE;
+	msg.hdr.rq = 1;
+
+	msg.buf = (uint8_t *)&req;
+	msg.len = sizeof(struct mctp_to_ipmi_sel_req);
+
+	if (set_iana(req.header.iana, sizeof(req.header.iana))) {
+		LOG_ERR("Set IANA fail");
+		return false;
+	}
+
+	req.header.netfn_lun = (NETFN_STORAGE_REQ << 2);
+	req.header.ipmi_cmd = CMD_STORAGE_ADD_SEL;
+	req.req_data.event.record_type = system_event_record;
+	req.req_data.event.gen_id[0] = (I2C_ADDR_BIC << 1);
+	req.req_data.event.evm_rev = evt_msg_version;
+
+	memcpy(&req.req_data.event.sensor_type, &sel_msg->sensor_type,
+	       sizeof(common_addsel_msg_t) - sizeof(uint8_t));
+
+	uint8_t resp_len = sizeof(struct mctp_to_ipmi_sel_resp);
+	uint8_t rbuf[resp_len];
+
+	if (!mctp_pldm_read(find_mctp_by_smbus(I2C_BUS_BMC), &msg, rbuf, resp_len)) {
+		LOG_ERR("mctp_pldm_read fail");
+		return false;
+	}
+
+	struct mctp_to_ipmi_sel_resp *resp = (struct mctp_to_ipmi_sel_resp *)rbuf;
+
+	if ((resp->header.completion_code != MCTP_SUCCESS) ||
+	    (resp->header.ipmi_comp_code != CC_SUCCESS)) {
+		LOG_ERR("Check reponse completion code fail %x %x", resp->header.completion_code, resp->header.ipmi_comp_code);
+		return false;
+	}
+
+	return true;
+}
+
+int pal_get_medium_type(uint8_t interface)
+{
+	int medium_type = -1;
+
+	switch(interface) {
+		case BMC_IPMB:
+		case MCTP:
+		case PLDM:
+		medium_type = MCTP_MEDIUM_TYPE_SMBUS;
+		break;
+		default:
+		medium_type = -1;
+		break;
+	}
+
+	return medium_type;
+}
+
+
+int pal_get_target(uint8_t interface)
+{
+	int target = -1;
+
+	switch(interface) {
+		case BMC_IPMB:
+		case MCTP:
+		case PLDM:
+		target = I2C_BUS_BMC;
+		break;
+		default:
+		target = -1;
+		break;
+	}
+
+	return target;
+}
+
+mctp *pal_get_mctp(uint8_t mctp_medium_type, uint8_t bus)
+{
+	switch (mctp_medium_type) {
+		case MCTP_MEDIUM_TYPE_SMBUS:
+		return find_mctp_by_smbus(bus);
+		default:
+		return NULL;
+	}
+}
+
+void plat_mctp_init(void)
+{
+	int ret = 0;
+
+	/* init the mctp/pldm instance */
+	for (uint8_t i = 0; i < ARRAY_SIZE(smbus_port); i++) {
+		mctp_smbus_port *p = smbus_port + i;
+
+		p->mctp_inst = mctp_init();
+		if (!p->mctp_inst) {
+			LOG_ERR("mctp_init failed!!");
+			continue;
+		}
+
+		uint8_t rc = mctp_set_medium_configure(p->mctp_inst, MCTP_MEDIUM_TYPE_SMBUS, p->conf);
+		if (rc != MCTP_SUCCESS) {
+			LOG_INF("mctp set medium configure failed");
+		}
+
+		mctp_reg_endpoint_resolve_func(p->mctp_inst, get_mctp_route_info);
+
+		mctp_reg_msg_rx_func(p->mctp_inst, mctp_msg_recv);
+
+		ret = mctp_start(p->mctp_inst);
```

Comment:
Better to print some error if start fails?

Reply:
The mctp_start will print error itself.
https://github.com/facebook/OpenBIC/blob/main/common/service/mctp/mctp.c#L481